### PR TITLE
Feature/mock dataset objects

### PR DESCRIPTION
**Pull request by alec-deason**
This PR has the new component management system that will be used for the data artifact changes that are coming soon. These changes are self contained and backward compatible.

New features:
- Components are read from config and initialized by a new ComponentManager class which can be replaced with a specialized subclass by specifying a different type in the config.
- The ComponentManager class tracks which components are being loaded which affords systems which need to be able to query the component list early in the initialization process (for example data artifact builders or risk factor implementations that are aware at setup time of what causes are being modeled (which would be future work))
- Adds the concept of DatasetManager, which is a dummy object in this version but is used by the data artifact implementation.
-  The ComponentManager can execute custom code in response to directives in component descriptions or placeholders in component object attributes. This is used by the data artifact implementation to determine which data to load out of the artifact during simulation setup and what data to include when creating the artifact.

### CODE REVIEW

File: vivarium/framework/state_machine.py [hunk 1]
```diff
@@ -339,11 +336,27 @@ def _normalize_probabilities(self, outputs, probabilities):
...
+
+    def __iter__(self):
+        return iter(self.transitions)
+
+    def __len__(self):
+        return len(self.transitions)
+
     def __str__(self):
-        return repr(self)
+        return repr(self.transitions)
```
Comment (collijk) [inline vivarium/framework/state_machine.py]: Think you still want repr(self) here so that the transitions get passed through str.

Comment (alec-deason) [reply vivarium/framework/state_machine.py]: I've changed this from being a list subclass to storing the transitions on the attribute self.transitions. So I think this does what you think it should, unless you're suggesting that there is something in the TransitionSet itself that should be captured in the string.

Comment (collijk) [reply vivarium/framework/state_machine.py]: Oh, no, it's just that it feels inconsistent that `str(transition_set)` essentially returns `[repr(x) for x in transition_set]` while `repr(transition_set)` returns `[str(x) for x in transition_set]`.

File: vivarium/framework/components.py [hunk 2]
```diff
@@ -1,63 +1,296 @@
...
+
+        Parameters
+        ----------
+        components:
+          Components to register
+        """
+
+        self.components = components + self.components
+
+    def setup_components(self, builder):
```
[the remark this answers is not in the document]
Comment (collijk) [reply vivarium/framework/components.py]: **For Discussion**
Also, we really should think about how to make the component setup more reasonable.  It should happen in three distinct phases: Gather the roots of component trees, flatten the trees, perform the setup.  I know that's not the goal of this PR, but we should do it soon.

File: vivarium/framework/components.py [hunk 3]
```diff
@@ -1,63 +1,296 @@
...
+       The configuration to read from
+    """
+
+    def _process_level(level, prefix):
         component_list = []
-        for c in level:
-            if isinstance(c, dict):
-                for k, v in c.items():
-                    component_list.extend(process_level(v, prefix + [k]))
+        for child in level:
```
Comment (collijk) [inline vivarium/framework/components.py]: Thanks for words!

File: vivarium/framework/components.py [hunk 4]
```diff
@@ -1,63 +1,296 @@
...
+
+    Parameters
+    ----------
+    desc
+        The component definition
+    constructors
+        Dictionary of callables for creating argument objects
+    """
+
+    component, *args = ast.iter_child_nodes(list(ast.iter_child_nodes(
```
Comment (collijk) [inline vivarium/framework/components.py]: This is maybe useful to have as a standalone function.  It needs an intelligible name and you also use it a few times in the tests.

File: vivarium/framework/components.py [hunk 5]
```diff
@@ -1,63 +1,296 @@
...
+        list(ast.iter_child_nodes(ast.parse(desc)))[0]))[0])
+    component = _component_ast_to_path(component)
+    new_args = []
+    for arg in args:
+        parsed = False
```
Comment (collijk) [inline vivarium/framework/components.py]: **For discussion**
`continue`

File: vivarium/framework/components.py [hunk 6]
```diff
@@ -1,63 +1,296 @@
...
+                constructor = constructors.get(val.__class__)
+                if constructor:
+                    setattr(component, attr, constructor(val))
 
             # Establish the initial configuration
             if hasattr(component, 'configuration_defaults'):
-                config.read_dict(component.configuration_defaults, layer='component_configs', source=module_path)
+                config.read_dict(component.configuration_defaults, layer='component_configs', source=component)
 
             if call:
```
Comment (collijk) [inline vivarium/framework/components.py]: I also think it's worth revisiting the all components should be classes discussion.  I think that makes a lot of the code in this module much simpler.

File: vivarium/framework/engine.py [hunk 7]
```diff
@@ -1,12 +1,13 @@
 """The engine."""
+import argparse
```
Comment (collijk) [inline vivarium/framework/engine.py]: **For Discussion**
Make engine a class.

File: tests/framework/test_components.py [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+
+import pytest
+import yaml
+
+from vivarium.framework.components import (_import_by_path, load_component_manager, ComponentManager,
+                                           DummyDatasetManager, ComponentConfigError, _extract_component_list,
+                                           _component_ast_to_path, _parse_component, ParsingError, _prep_components)
+
+
+# Fiddle the path so we can import from this module
```
Comment (collijk) [inline tests/framework/test_components.py]: **For Discussion**
This weirdness.

File: tests/framework/test_components.py [hunk 9]
```diff
@@ -0,0 +1,238 @@
+import sys
+import os
+import ast
+from unittest.mock import patch, mock_open
```
Comment (collijk) [inline tests/framework/test_components.py]: **For Discussion**
Moving to pytest-mock

File: vivarium/framework/components.py [hunk 10]
```diff
@@ -1,63 +1,296 @@
...
+    pass
+
+
+class DummyDatasetManager:
+    """Placeholder implementation of the DatasetManager"""
+    def __init__(self):
+        self.constructors = {}
+
+
+def _import_by_path(path: str) -> Union[type, Callable]:
```
[the remark this answers is not in the document]
Comment (alec-deason) [reply vivarium/framework/components.py]: Hmm. You may be right that `Union[type, Callable]` is redundant. I gave up trying to use mypy to validate these declarations. In my head I'm going 'it's a class or a function' but you're right that a class is also a Callable.

File: vivarium/framework/components.py [hunk 11]
```diff
@@ -1,63 +1,296 @@
...
+    """Import a class or function given it's absolute path.
+
+    Parameters
+    ----------
+    path:
+      Absolute class to object to import
```
[the remark this answers is not in the document]
Comment (alec-deason) [reply vivarium/framework/components.py]: Well, that typo makes my docstring almost unreadable but I think in our case the paths will always be absolute though that being said the function will also work on relative paths so it should specify in the docstring.

File: vivarium/framework/components.py [hunk 12]
```diff
@@ -1,63 +1,296 @@
...
 
-def read_component_configuration(path):
-    if path.endswith('.yaml'):
-        with open(path) as f:
-            component_config = yaml.load(f)
-            return prepare_component_configuration(component_config, path=path)
+from vivarium import config
+
+
+class ComponentConfigError(Exception):
```
[the remark this answers is not in the document]
Comment (alec-deason) [reply vivarium/framework/components.py]: Yep

File: vivarium/framework/components.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (alec-deason) [reply vivarium/framework/components.py]: mypy agrees that type is a Callable so I'll use that.

[the remark this answers is not in the document]
Comment (collijk) [reply vivarium/framework/components.py]: Excellent.  I wasn't actually certain.

**Comment by alec-deason** [comment]
Yeah, I need to actually read the peps to understand how this stuff is
meant to be interpreted. But I haven't done it yet.

**Pull request merged**